### DISCUSSION TITLE
"Behind a Proxy" changed behaviour in the new version 0.109.0

### DISCUSSION THREAD

**Discussion by neugeeug** [fastapi/fastapi#10948] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
uvicorn rest.main:server --host 0.0.0.0 --port 8080 --reload --root-path /proxy

server = FastAPI()

# separate spec at /v1/docs
server.mount("/v1", app_v1)
# separate spec at /v2/docs
server.mount("/v2", app_v2)

#where app_v1, app_v2 is like below
app_v1 = FastAPI(openapi_tags=tags_metadata, title="Pet store")
```

### Description

When I run the uvicorn with the `--root-path` option for the above setup, the server is expecting that I include the defined root path in every request e.g:
 http://localhost:8080/v2/docs is not working (404)
 http://localhost:8080/proxy/v2/docs is working

IN version 0.108.0 It works as described here https://fastapi.tiangolo.com/advanced/behind-a-proxy/

### Operating System

Linux

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

3.11.5

**Comment by abramjstamper** [comment]
I updated yesterday to 0.109.0 and ran into the same issue today.

```
uvicorn app:main --host 0.0.0.0 --port 80 --root-path "/test"

router = APIRouter(prefix="/test")
@router.get("/")
async def test():
    return "will 404"

@router.get("/test123")
async def test():
    return "will 404"

router_alt = APIRouter()
@router.get("/test")
async def test():
    return "will 404"

app = FastAPI()
app.include_router(router)
app.include_router(router_alt)
```

Even though the requests aren't working correctly, they render correctly in the swagger docs. It seems anything that has the same leading characters as `--root-path` fails. Setting `path_prefix` instead of `--root-path` doesn't seem to make any difference either.

I believe this regression was introduced in starlette v0.35.0 with [this PR](https://github.com/encode/starlette/pull/2400).

**Reply by mawiegand** [reply]
I also had an issue with it as described here: https://github.com/tiangolo/fastapi/pull/10798#issuecomment-1878611386

**Comment by andy-5** [comment]
I have exactly the same issue with 0.109.0 (and starlette 0.35.1). It seems `--root-path` together with sub-applications is broken.

To reproduce, i used the simple example from https://fastapi.tiangolo.com/advanced/sub-applications/ and started it with

    uvicorn main:app --root-path /api/v1

as explained in https://fastapi.tiangolo.com/advanced/behind-a-proxy/ .

Accessing `http://127.0.0.1:9999/api/v1/subapi/sub` through the example Traefik setup, or directly accessing `http://127.0.0.1:8000/subapi/sub` results in `404`.

It works fine with FastAPI `0.108.0`.

**Comment by tiangolo** [comment, marked as answer]
Thank you! Indeed, this is an issue, it's because recently Starlette was updated to more strictly follow the ASGI specification, and now, since versions 0.34.0 and 0.35.0 it is more *correct*, nevertheless, Uvicorn needs the same/equivalent change.

Uvicorn has an ongoing PR here: https://github.com/tiangolo/fastapi/issues/10978

Once that is taken and there's a new release including it, that will be solved. As a temporary workaround, you can pin FastAPI to a previous version (that depended on a previous version of Starlette), or you might try Hypercorn.

I created an issue to track this here: https://github.com/tiangolo/fastapi/issues/10978

**Reply by neugeeug** [reply]
Thank you !
